### PR TITLE
Fix concurrency error in registry's pull-through

### PR DESCRIPTION
**Pull request by miminar**
Avoid the use of not thread-safe structures in the pull-through code. The mirroring code spawns a new go-routine that needs to have its own instances of thread-dangerous data.

Resolves: #14434

### CODE REVIEW

**Comment by miminar** [comment]
@dmage, @legionus PTAL

**Comment by miminar** [comment]
[test][testextended][extended:core(\[registry\])]

File: pkg/dockerregistry/server/pullthroughblobstore.go [hunk 1]
```diff
@@ -179,8 +176,40 @@ func (pbs *pullthroughBlobStore) copyContent(store BlobGetterService, ctx contex
 	return desc, nil
 }
 
+// storeLocalInBackground spawns a separate thread to copy the remote blob from the remote registry to the
+// local blob store.
+// The function assumes that localBlobStore is thread-safe.
+func storeLocalInBackground(ctx context.Context, repo *repository, localBlobStore distribution.BlobStore, dgst digest.Digest) {
+	// leave only the essential entries in the context
+	newCtx := context.WithValue(context.Background(), "instance.id", ctx.Value("instance.id"))
```
Comment (dmage) [inline pkg/dockerregistry/server/pullthroughblobstore.go]: docker/distribution/context already provides "instance.id".
I would like to preserve "trace.id", although we don't use distribution/context.WithTrace yet.

Comment (miminar) [inline pkg/dockerregistry/server/pullthroughblobstore.go]: > docker/distribution/context.Background already provides "instance.id".

My intention was to make it possible to associate this job with the original request, but a better way to do that would probably be based on `http.request.id` instead, which is already logged by the logger.

I'll ditch the `instance.id` and `instance.jobid` and let the context generate new `instance.id` instead.

> I would like to preserve "trace.id", although we don't use distribution/context.WithTrace yet.

We don't use it yet so I don't see a point in preserving it. Once we start to use it, this function will set the `trace.*` attributes on its own anyway.

Comment (dmage) [inline pkg/dockerregistry/server/pullthroughblobstore.go]: instance.id is an id of the instance, it is the same across all requests. It is changed only when the application is restarted.

**Comment by miminar** [comment]
There's an extended test failure in @mfojtik's signing workflow test:

```
• Failure [310.874 seconds]
[imageapis][registry] image signature workflow
/openshifttmp/openshift/build-rpm-release/tito/rpmbuild-originuPn0fV/BUILD/origin-3.6.0/_output/local/go/src/github.com/openshift/origin/test/extended/registry/signature.go:110
  can push a signed image to openshift registry and verify it [It]
  /openshifttmp/openshift/build-rpm-release/tito/rpmbuild-originuPn0fV/BUILD/origin-3.6.0/_output/local/go/src/github.com/openshift/origin/test/extended/registry/signature.go:109

  Expected error:
      <*errors.errorString | 0xc4215b0030>: {
          s: "timed out while waiting of an image stream tag extended-test-registry-signing-ktm0k-v5tjz/signer:latest",
      }
      timed out while waiting of an image stream tag extended-test-registry-signing-ktm0k-v5tjz/signer:latest
  not to have occurred

  /openshifttmp/openshift/build-rpm-release/tito/rpmbuild-originuPn0fV/BUILD/origin-3.6.0/_output/local/go/src/github.com/openshift/origin/test/extended/registry/signature.go:30
```

It fails every time. Not sure if my changes are to be blamed. I'll try to reproduce locally.

**Comment by miminar** [comment]
Flake #14496

**Comment by miminar** [comment]
@soltysh merge please.

**Comment by soltysh** [comment]
Got @dmage approval on IRC.

[merge]

**Pull request merged**